### PR TITLE
Basic RSS support.

### PR DESCRIPTION
**Pull request by qnikst**

### CODE REVIEW

File: app/Main.hs [hunk 1]
```diff
@@ -22,16 +23,18 @@ main = do
     let configFile = case args of
             [cnf] -> cnf
             _ -> error usage
-    run configFile repostUpdates
+    run configFile
   where
```
Comment (qnikst) [inline app/Main.hs:27]: Not sure that I like this change, but I was lazy to add many instances for `RssT`, `DiscourseT` and `GitterT`.

File: src/RSS.hs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+extractItems :: Document -> [Item]
+extractItems xml = concat $
+      (\x -> channelItem (channelTitle x) x)
+          <$> xml ^.. root ./ el "channel"
+  where
+    channelTitle x = x ^. el "channel" ./ el "title" . text
+    channelItem t x = (Item <$> itemTitle <*> itemLink <*> pure t)
+      <$> x ^.. el "channel" ./ el "item"
+    itemTitle x = x ^. el "item" ./ el "title" . text
+    itemLink x  = x ^. el "item" ./ el "link" . text
```
Comment (qnikst) [inline src/RSS.hs:74]: Here, I had to use `el "channel"`/`el "item"` in parent and child lens, I think that it's not correct, but I had not found better way with `lens-xml` package.

File: app/Main.hs (hunk 1 above)
Comment (cblp) [inline app/Main.hs:27]: I didn't understand the point of this change. In previous version, `run` was responsible for unpacking the transformer stack down to the base monad. In your version, it runs the whole transformed monad. Both semantics are correct, but they are simply two different semantics.

Comment (cblp) [inline app/Main.hs:27]: I'll clean the code in this branch, according to my own vision, and then I'll merge.

File: src/RSS/Cache.hs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import Data.Aeson
+import Data.Text (Text)
+
+-- | Cache of the loaded links.
+newtype RSSCache = RSSCache (HashMap FeedUrl [Text]) deriving (Show)
+
+emptyCache :: RSSCache
+emptyCache = RSSCache HM.empty
+
+-- Manua instances reuse internal aeson structure
```
Comment (cblp) [inline src/RSS/Cache.hs]: typo

File: app/Main.hs (hunk 1 above)
Comment (cblp) [inline app/Main.hs:27]: Forget that. It seems working. I'll merge this, and then refactor.

File: config.json [hunk 4]
```diff
@@ -8,5 +8,10 @@
       "uri": ["cblp", "discourse-to-gitter"]
     },
     "tokenFile": "../../felix/config/Gitter/HaskellCurry-token"
-  }
+  },
+  "rssCache": "cacherss",
+  "rss": {
+      "feeds": ["http://bananasandlenses.net/feed.xml"]
+    , "prepend": false
```
Comment (cblp) [inline config.json]: This isn't used. What is this?

File: src/RSS.hs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import RSS.Cache
+import Gitter.Monad
+
+-- global
+import Network.Wreq
+import Data.Foldable
+import Data.Monoid
+import qualified Data.Set as Set
+import qualified Data.Text as Text
+import qualified Data.HashMap.Strict as HM
```
Comment (cblp) [inline src/RSS.hs]: Style note: don't use excessive abbreviations in imports. `HashMap` is short enough.

File: src/RSS.hs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import Text.XML.Lens
+import Text.XML
+import Control.Monad.Reader
+import Control.Monad.Catch
+
+-- | Basic RSS monad.
+newtype RssT m a = RssT { _runRssT :: ReaderT RSS m a }
```
Comment (cblp) [inline src/RSS.hs]: Style note: `_runRssT` is not used. Why not omit it?

File: src/RSS.hs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    channelItem t x = (Item <$> itemTitle <*> itemLink <*> pure t)
+      <$> x ^.. el "channel" ./ el "item"
+    itemTitle x = x ^. el "item" ./ el "title" . text
+    itemLink x  = x ^. el "item" ./ el "link" . text
+
+-- | Update cache, if feed does not exist in it, then add that feed.
+initializeFeeds :: RSSCache -> [(FeedUrl,[Item])] -> RSSCache
+initializeFeeds = foldl' initialize
+  where
+    initialize c@(RSSCache h) (u,l) = case HM.lookup u h of
```
Comment (cblp) [inline src/RSS.hs]: Style note: don't use obscure names if you can use descriptive ones. What is `c`? What is `u`? `l`? `h`?

**Comment by cblp** [comment]
@qnikst There's a problem with algorithm. I think it's more user-friendly to repost one last message when a feed first added. This logic is implemented in `Lib.detectNewTopics`. Please make RSS code use this function or implement the same logic. Or we can discuss an alternative behaviour.

**Comment by cblp** [comment]
Another problem is with cache size. Will it grow without limits?

**Comment by cblp** [comment]
I didn't tested, but `ToJSON RSSCache` and `FromJSON RSSCache` instances don't look inverse. `FromJSON` uses "rss-cache", but `ToJSON` doesn't.

**Comment by cblp** [comment]
Why do you need solid RSSCache datatype? Why not just type alias?

File: app/Main.hs (hunk 1 above)
Comment (qnikst) [inline app/Main.hs:27]: The basic problem that I had to do a lot of unrelated changes otherwise. It's not possible to use two FileCache layer in one monad stack, because they have different types. As a result MonadStack should be reordered as:

```
runGitterT $ flip runReaderT config $ do
    runDiscourceT $ runFileCacheT ...
    runRssT $ runFileCacheT ..
```

so but it will require changes in how we pass arguments to `run`, and other changes in existing instances.

File: config.json (hunk 4 above)
Comment (qnikst) [inline config.json]: Thanks, forgot to remove.

File: src/RSS.hs (hunk 5 above)
Comment (qnikst) [inline src/RSS.hs]: Ok.

File: src/RSS.hs (hunk 6 above)
Comment (qnikst) [inline src/RSS.hs]: Ok.

File: src/RSS.hs (hunk 7 above)
Comment (qnikst) [inline src/RSS.hs]: Sorry, I don't see any problems with short names here, however if you insist on adding fancy long or it does not match coding style if this project, I'll change naming.

File: src/RSS/Cache.hs (hunk 3 above)
Comment (qnikst) [inline src/RSS/Cache.hs]: Thanks

**Comment by qnikst** [comment]
There is a core difference between Discource module that support only one source, and RSS that support many feeds at once. Returning only one message instead of all _new _messages looks like bad idea. We could lose notifications about new post in a random feed, in that case. This is ok in case of discource, because you will read other messages anyway, but it's not a case if we add more than one feed. So exactly the same algorithm completely breaks the purpose of notification system. 
There are few ideas how we could improve notifications system, though
1. We could batch all messages into one (or one per feed). Then result will depend on how often you'll run this program. For example most of the feeds I have are updated each 2 hour, and I almost never receive more than one message, so I think it's more theoretical problem than real one
2. However we could add additional protection for the case if user completely change RSS settings so all messages will be rendered as a new messages. We could check in there is more than 2 messages in the feed and output only last message in that case.

Another difference is that on the contrary to Discource where there is a nice well defined API, RSS feeds are generated by user libraries and could contain bugs of all sorts, such are different formats of timestamp, add messages in wrong order, write wrong information in guid fields.

Cache is unlimited if it's a problem for you, I could keep only last 10 messages in it. This will be enough to catch problems 

There is no problem with RSSCache JSON instance, "rss-cache" is used only for error reporting. It's implemented as a newtype in order to write JSON instance for that type.

If you  agree with my comment, I could address all review comments. And update this PR.

File: src/RSS.hs (hunk 7 above)
Comment (cblp) [inline src/RSS.hs]: Not a big problem. I'll rename them later.

Comment (qnikst) [inline src/RSS.hs]: I could agree with `u` for `url`, and `l` for `link`, but I suppose it's longer to type "what is `c`? `h`?" than to find (not sure if word "find" is right here) the answer in the code, and this will not be different if I use `cache` or `hash` instead of of `c` and `h`. Really I'm interested in how other people understand code and what could be hard/unexpected for them.
Btw, does it apply to short names in `updateFeeds`?

Comment (cblp) [inline src/RSS.hs]: It is faster for one developer to type `cache` or `hash` than for many other developers to try to find their sense in the code around.

Comment (cblp) [inline src/RSS.hs]: Yes, it applies to the other functions.

Not every one-letter variable is bad, only non-obvios. For example it's obvoius that `f` is for function. But `z` for cache?

**Comment by cblp** [comment]
> There is a core difference between Discource module that support only one source, and RSS that support many feeds at once.

There is no core difference. We can change the code easily if we want to support more Discourse sites. Discourse or RSS — just kinds of the same entity — the news source.

**Comment by cblp** [comment]
> Returning only one message instead of all _new _messages looks like bad idea. We could lose notifications about new post in a random feed, in that case.

Yes! And I never return one message instead of _all new messages_. See `Lib.detectNewTopics` please. If we can determine exactly which items are new, then we return them.

But when feed is just added, and we received some items, are they new or old? There are several choices.
1. Consider all items new. Maybe the most correct choice, but due to its verbosity, it is not usable.
2. Consider all items old (your choice). How can I see if it works?
3. Consider one last item new (my choice). Two birds with one stone! The program instantly shows the user that it works correctly. And it is not over-verbose.

**Comment by qnikst** [comment]
> Yes! And I never return one message instead of all new messages. See Lib.detectNewTopics please. If we can determine exactly which items are new, then we return them.

But my my code does exactly the same, I don't understand your concern.

> But when feed is just added, and we received some items, are they new or old? 

if feed is just added then all messages are concidered as old, should I add documentation to `initializeFeeds` to make this explicit?

**Comment by qnikst** [comment]
Ah.. I see, ok, 3. is fine with me.

**Comment by cblp** [comment]
> 1. We could batch all messages into one (or one per feed). Then result will depend on how often you'll run this program. For example most of the feeds I have are updated each 2 hour, and I almost never receive more than one message, so I think it's more theoretical problem than real one

Right now the program is beeing run every minute. I see no problem sending every piece of news as a separate message.

> 1. However we could add additional protection for the case if user completely change RSS settings so all messages will be rendered as a new messages. We could check in there is more than 2 messages in the feed and output only last message in that case.

Do you mean the RSS feed's author? It is very rare to change RSS completely, and in that case it's better to show as many news as possible. I don't think we should prepare much for this case.

More interesting case is deleting a message from a feed by its author or increasing the length of the feed. In this case we can get some really old messages.

> Cache is unlimited if it's a problem for you, I could keep only last 10 messages in it. This will be enough to catch problems

And if you set the fixed cache size, your algirithm will detect those old messages as new. My `Lib.detectNewTopics` won't, if RSS feed has proper timestamps. We can't rely on timestamps in real-life RSS feeds, and I'm sad about it.

**Comment by cblp** [comment]
What about declining RSS feeds without proper timestamps?

**Comment by cblp** [comment]
> Ah.. I see, ok, 3. is fine with me.

So will you change the code or will you leave it to me?

**Comment by qnikst** [comment]
Proper == one we could parse? I tested 3 different feed and all of them had different format of timestamps in messages, this I could test more feeds, and then choose parsers. If timestamp will fail to be parsed by all of them then drop it.

**Comment by qnikst** [comment]
@cblp yes, I'll change the code.

**Comment by cblp** [comment]
Oh, different timestamps format...

The other way I see is to keep in cache fixed but great number of items, knowignly greater than the typical feed has. Say, 1000.

**Comment by cblp** [comment]
Don't change the cache policy in this branch, I will refactor it separately.

**Comment by cblp** [comment]
One little thing. It's often better to identify feed items by `guid`, falling back to `link`.

**Comment by qnikst** [comment]
Pushed style fixes, and update of initial rss load.

Not done: guid -> link identification, I'm not sure if I can trust guid, i.e. if rss is generated by some library for the static will will it always generate the same guid for the same post? Also I could add data parsing, all possible parsers could we written to config, so it will not require recompilation if any parser will be missing.

**Comment by cblp** [comment]
Merged!
